### PR TITLE
Feature/refactor filters

### PR DESCRIPTION
**Pull request by gnusi**
* refactor filters
* refactor stat collectors
* avoid creation of automaton (for automaton based filters) for every segment
* teach phrase filter to honor boost from referenced filters
* switch to c++17

### CODE REVIEW

File: core/search/granular_range_filter.cpp [hunk 1]
```diff
@@ -334,11 +333,11 @@ void collect_terms_within(
...
 
   if (min_term_inclusive && !min_term.empty()) {
-    auto& exact_min_term = min_term.begin()->second;
-    bool single_term = !max_term.empty() && exact_min_term == max_term.begin()->second;
+    auto& exact_min_term = min_term.front();
+    bool single_term = !max_term.empty() && exact_min_term == max_term.front();
 
     if ((!single_term || max_term_inclusive)
-        && exact_min_term > max_term.begin()->second) {
+        && exact_min_term > max_term.front()) {
```
Comment (Dronplane) [inline core/search/granular_range_filter.cpp:340]: Could max_term be empty there ? Than single_term is false and max_term.front() will trigger UB.
Suggest to add max_term.empty() check

File: core/search/granular_range_filter.cpp [hunk 2]
```diff
@@ -352,7 +351,7 @@ void collect_terms_within(
   // align the min_term granularity level with max_term granularity level
   // ...........................................................................
 
-  auto* exact_min_term = min_term.empty() ? nullptr : &(min_term.begin()->second);
+  auto* exact_min_term = min_term.empty() ? nullptr : &(min_term.front());
```
Comment (Dronplane) [inline core/search/granular_range_filter.cpp:354]: Is it redundant to check min_term.empty() after check in line 322 ?

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/iresearch_s-pr/90/

File: core/search/term_query.hpp [hunk 3]
```diff
@@ -23,54 +23,31 @@
...
+class term_query final : public filter::prepared {
  public:
-  typedef states_cache<reader_term_state> states_t;
+  //////////////////////////////////////////////////////////////////////////////
+  /// @class term_state
+  /// @brief cached per reader term state
+  //////////////////////////////////////////////////////////////////////////////
+  struct term_state {
+    term_state() = default;
+    term_state(term_state&& rhs) = default;
```
Comment (iurii-i-popov) [inline core/search/term_query.hpp:42]: Did you delete copy constructor and assignment operators intentionally?

File: core/utils/hash_utils.hpp [hunk 4]
```diff
@@ -38,7 +38,7 @@ FORCE_INLINE size_t hash_combine(size_t seed, size_t v) noexcept {
 }
 
 template<typename T>
-FORCE_INLINE size_t hash_combine(size_t seed, T const& v) {
+FORCE_INLINE size_t hash_combine(size_t seed, T const& v) noexcept(noexcept(std::hash<T>()(v))) {
```
Comment (iurii-i-popov) [inline core/utils/hash_utils.hpp:41]: Why did you add a wrapping noexcept()?

File: tests/search/term_filter_tests.cpp [hunk 5]
```diff
@@ -656,12 +661,11 @@ TEST(by_term_test, boost) {
...
   {
     irs::boost_t boost = 1.5f;
-    irs::by_term q;
-    q.field("field").term("term");
+    irs::by_term q = make_filter("field", "term");
     q.boost(boost);
 
     auto prepared = q.prepare(irs::sub_reader::empty());
-    ASSERT_EQ(boost, prepared->boost());
+
```
Comment (iurii-i-popov) [inline tests/search/term_filter_tests.cpp:668]: Where is the test?

File: core/search/term_query.hpp (hunk 3 above)
Comment (gnusi) [reply core/search/term_query.hpp:42]: in fact it's implicitly deleted by `std::unique_ptr<...>` (see `seek_term_iterator::cookie_ptr cookie`) member

File: core/utils/hash_utils.hpp (hunk 4 above)
Comment (gnusi) [reply core/utils/hash_utils.hpp:41]: because it's `noexcept` operator is called from within `noexcept` specifier, otherwise it won't compile

File: tests/search/term_filter_tests.cpp (hunk 5 above)
Comment (gnusi) [reply tests/search/term_filter_tests.cpp:668]: good catch

File: core/search/term_query.hpp (hunk 3 above)
Comment (iurii-i-popov) [reply core/search/term_query.hpp:42]: Compare with `multiterm_state::term_state`. You have left `term_state& operator=(term_state&& rhs) = default;`

File: core/search/wildcard_filter.hpp [hunk 6]
```diff
@@ -23,19 +23,58 @@
...
+/// @brief options for wildcard filter
+////////////////////////////////////////////////////////////////////////////////
+struct IRESEARCH_API by_wildcard_options : by_wildcard_filter_options {
+  using filter_type = by_wildcard;
+  using filter_options = by_wildcard_filter_options;
+
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief the maximum number of most frequent terms to consider for scoring
+  //////////////////////////////////////////////////////////////////////////////
+  size_t scored_terms_limit{1024};
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/wildcard_filter.hpp:60]: I suggest to leave it as it is now for 2 reasons:
* after refactoring default value became more explicit
* I plan to generalize concept of such execution options in future

File: core/search/range_filter.hpp [hunk 7]
```diff
@@ -23,63 +23,80 @@
...
+/// @brief options for prefix filter
+////////////////////////////////////////////////////////////////////////////////
+struct IRESEARCH_API by_range_options : by_range_filter_options {
+  using filter_type = by_range;
+  using filter_options = by_range_filter_options;
+
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief the maximum number of most frequent terms to consider for scoring
+  //////////////////////////////////////////////////////////////////////////////
+  size_t scored_terms_limit{1024};
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/range_filter.hpp:63]: I suggest to leave it as it is now for 2 reasons:
* after refactoring default value became more explicit
* I plan to generalize concept of such execution options in future
